### PR TITLE
feat: use `ParOps.mapValues` in `Lowering`

### PR DESCRIPTION
**Pull request by sockmaster27**
Related to #2974

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/Lowering.scala [hunk 1]
```diff
@@ -142,27 +142,23 @@ object Lowering {
...
-    val effects = ParOps.parMap(root.effects.values)((e: TypedAst.Effect) => visitEffect(e)(root, flix))
-    val aliases = ParOps.parMap(root.typeAliases.values)((a: TypedAst.TypeAlias) => visitTypeAlias(a)(root, flix))
-
-    val newDefs = defs.map(kv => kv.sym -> kv).toMap
-    val newSigs = sigs.map(kv => kv.sym -> kv).toMap
-    val newInstances = instances.map(kv => kv.head.clazz.sym -> kv).toMap
-    val newEnums = (enums ++ restrictableEnums).map(kv => kv.sym -> kv).toMap
-    val newEffects = effects.map(kv => kv.sym -> kv).toMap
-    val newAliases = aliases.map(kv => kv.sym -> kv).toMap
+    val defs = ParOps.mapValues(root.defs)((d: TypedAst.Def) => visitDef(d)(root, flix))
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Lowering.scala:145]: Is it possible to simply write `ParOps.mapValues(root.defs)(visitDef)` or is Scala's type inference too weak?

**Comment by magnus-madsen** [comment]
Updated the title.

**Comment by magnus-madsen** [comment]
Whats the impact on --Xbenchmark-throughput?

**Comment by sockmaster27** [comment]
===== Before =====
Throughput (best): 20.187 lines/sec (with 16 threads.)

  min:  4.911, max: 20.187, avg: 16.520, median: 18.732

  The highest throughput was in iteration: 5 (out of 10).
  The ratio between the best and worst iteration was: 4,1x.

Finished 10 iterations on 66.126 lines of code in 47 seconds.

Process finished with exit code 0

0






===== After =====
Throughput (best): 21.896 lines/sec (with 16 threads.)

  min:  4.013, max: 21.896, avg: 17.718, median: 20.437

  The highest throughput was in iteration: 6 (out of 10).
  The ratio between the best and worst iteration was: 5,5x.

Finished 10 iterations on 66.126 lines of code in 48 seconds.

Process finished with exit code 0

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
Will merge when tests complete. Will you keep track in the ticket of which phases has been updated?

**Comment by sockmaster27** [comment]
> Will merge when tests complete. Will you keep track in the ticket of which phases has been updated?

I think I'll just do one more PR when you give me that tip on what to do about `Validation.sequence` :)

**Comment by magnus-madsen** [comment]
> > Will merge when tests complete. Will you keep track in the ticket of which phases has been updated?
> 
> I think I'll just do one more PR when you give me that tip on what to do about `Validation.sequence` :)

I have forgotten everything. Where/what's the question?

**Comment by sockmaster27** [comment]
> > > Will merge when tests complete. Will you keep track in the ticket of which phases has been updated?
> > 
> > 
> > I think I'll just do one more PR when you give me that tip on what to do about `Validation.sequence` :)
> 
> I have forgotten everything. Where/what's the question?

This pattern is very common and I just can't think of the clever rewrite:
```
val a = ParOps.parMap(...)(...)
val b = Validation.sequence(a)
mapN(b, ...) {
    case (c, ...) => 
        (c.toMap, ...)
```

**Comment by magnus-madsen** [comment]
> > > > Will merge when tests complete. Will you keep track in the ticket of which phases has been updated?
> > > 
> > > 
> > > I think I'll just do one more PR when you give me that tip on what to do about `Validation.sequence` :)
> > 
> > 
> > I have forgotten everything. Where/what's the question?
> 
> This pattern is very common and I just can't think of the clever rewrite:
> 
> ```
> val a = ParOps.parMap(...)(...)
> val b = Validation.sequence(a)
> mapN(b, ...) {
>     case (c, ...) => 
>         (c.toMap, ...)
> }
> ```

I think it is impossible to eliminate sequence here -- because it fundamentally collapses a list of validation things into a validation with a list of things and lists of errors in it.

**Comment by magnus-madsen** [comment]
Of course we can have a helper function inside ParOps to do that, but the fundamental structure will be there.

**Pull request merged**